src/Interpreters/ErrorLog.cpp

```diff
@@ -0,0 +1,127 @@
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/ErrorLog.h>
+#include <base/getFQDNOrHostName.h>
+#include <Common/DateLUTImpl.h>
+#include <Common/ThreadPool.h>
+#include <Common/ErrorCodes.h>
+
+#include <vector>
+
+namespace DB
+{
+
+ColumnsDescription ErrorLogElement::getColumnsDescription()
+{
+    return ColumnsDescription {
+        {"hostname", std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()), "Hostname of the server executing the query."},
+        {"event_date", std::make_shared<DataTypeDate>(), "Event date."},
+        {"event_time", std::make_shared<DataTypeDateTime>(), "Event time."},
+        {"code", std::make_shared<DataTypeInt32>(), "Error code}"},
+        {"error", std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()), "Error name."},
+        {"value", std::make_shared<DataTypeUInt64>(), "Number of errors happened in time interval."},
+        {"remote", std::make_shared<DataTypeUInt8>(), "Remote exception (i.e. received during one of the distributed queries)."}
+    };
+}
+
+void ErrorLogElement::appendToBlock(MutableColumns & columns) const
+{
+    size_t column_idx = 0;
+
+    columns[column_idx++]->insert(getFQDNOrHostName());
+    columns[column_idx++]->insert(DateLUT::instance().toDayNum(event_time).toUnderType());
+    columns[column_idx++]->insert(event_time);
+    columns[column_idx++]->insert(code);
+    columns[column_idx++]->insert(ErrorCodes::getName(code));
+    columns[column_idx++]->insert(value);
+    columns[column_idx++]->insert(remote);
+}
+
+void ErrorLog::startCollectError(size_t collect_interval_milliseconds_)
+{
+    collect_interval_milliseconds = collect_interval_milliseconds_;
+    is_shutdown_error_thread = false;
+    flush_thread = std::make_unique<ThreadFromGlobalPool>([this] { threadFunction(); });
+}
+
+
+void ErrorLog::stopCollectError()
+{
+    bool old_val = false;
+    if (!is_shutdown_error_thread.compare_exchange_strong(old_val, true))
+        return;
+    if (flush_thread)
+        flush_thread->join();
+}
+
+
+void ErrorLog::shutdown()
+{
+    stopCollectError();
+    stopFlushThread();
+}
```

Comment:
This code is essentially the same as the one from `MetricLog`. I thought about consolidating in a single class this part so that both `MetricLog` and `ErrorLog` inherit from it, but I'm not sure it's worth it.

If we could consolidate also the code for the `threadFunction` itself it'd be even better, but it's not that easy to do if we want to have some state such as `previous_values` across loop iterations.

Reply:
I think this is Ok.

Reply:
I ended up consolidating the code into a `PeriodicLog` because I see there are [other issues for other periodic logs](https://github.com/ClickHouse/ClickHouse/issues/40012). So, I preferred to bite the bullet now while it's fresh